### ISSUE TITLE
 Incorrect DAG Run ID opened when clicking on run_id link from DAG run list

### ISSUE THREAD

**Issue by hkc-8010** [apache/airflow#41406]
### Apache Airflow version

2.9.3

### What happened?

When I try to open an older DAG run by clicking on the run_id link from the DAG run list, it incorrectly opens the DAG run with a different run_id. This issue makes it difficult to view the details of specific older DAG runs.

### What you think should happen instead?

Clicking on a run_id link from the dag run list should open the DAG run corresponding to that run_id.

### How to reproduce

- Navigate to the DAG runs list.
- Select a specific DAG to view its run history.
- Click on the run_id link of an older DAG run in the list.
- Notice that the page loads a DAG run with an incorrect run_id.

### Operating System

Linux ad25902d8cef 6.7.12-orbstack-00201-g2ddb8f197a46 #1 SMP Tue May 21 04:38:26 UTC 2024 aarch64 GNU/Linux

### Deployment

Astronomer

### Deployment details

Start an Airflow instance using the AstroCLI command `astro dev start`, run an example dag multiple times, and try accessing dag_run older than 25 from the dag runs list page of the Airflow UI.

### Anything else?

Click on the older run_id link from the dag run list

<img width="1394" alt="Screenshot 2024-08-12 at 8 14 16 PM" src="https://github.com/user-attachments/assets/bf2765d7-99a0-46d0-b2b9-eea8af72314e">

It will open the graph view. Click on any task in the graph view, and it will show the details for the incorrect dag run_id.

<img width="1428" alt="Screenshot 2024-08-12 at 8 14 27 PM" src="https://github.com/user-attachments/assets/73bf497f-a4ec-4e71-8d65-00f421bc437e">

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by tirkarthi** [comment]
Looks like a duplicate of https://github.com/apache/airflow/issues/40214

**Comment by RNHTTR** [comment]
> Looks like a duplicate of #40214

That Issue appears to be solved in 2.9.3. @hkc-8010 can you please confirm that this is occurring on 2.9.3?

**Comment by hkc-8010** [comment]
It appears that the issue has been resolved for the dag_id link, which points to the Grid view. However, the run_id link, pointing to the Graph view, is still incorrect. I was able to replicate this issue on version 2.9.3.

**Comment by tirkarthi** [comment]
@hkc-8010  Does run_id in the details tab match the expected value and the query string in the url? Some changes were done in https://github.com/apache/airflow/issues/35946

**Comment by hkc-8010** [comment]
@tirkarthi - Here is hyperlink attached to the run_id in the Dag run list - [manual__2024-08-09T08:42:25.524227+00:00](http://localhost:8080/dags/example_astronauts/grid?dag_run_id=manual__2024-08-09T08%3A42%3A25.524227%2B00%3A00&tab=graph)

The hyperlink appears correct, but when it opens the Graph view and I click on a task, it displays an incorrect `run_id`—specifically, the `run_id` from the previous run instead of the desired one.

It seems the hyperlink is updating the calendar tab to the exact `run_id` value, which causes the Graph or Grid view to display only historical DAG runs, excluding the desired DAG run.
 
<img width="828" alt="image" src="https://github.com/user-attachments/assets/06c1be20-d612-4a96-8bf3-b93d4dddeae0">

**Comment by hkc-8010** [comment]
I believe this [PR](https://github.com/apache/airflow/pull/40764) might resolve the issue. I'll test it on version 2.10.0 and will update you here. Thanks!

**Comment by hkc-8010** [comment]
I just verified on 2.10. Still facing the same issue.

**Comment by karenbraganz** [comment]
Hem and I tested this out in Airflow 2.10 (Astro runtime 12.0.0) and noticed that this bug still exists but only for certain types of runs (when the run_id timestamp is not at the 0th second of the minute).

- When testing this out with runs scheduled to run every minute, I did not see this bug. I was able to go to Browse -> DAG Runs, find a DAG run older than the 25th record, click on the run ID, and open the correct DAG run page.

- However, when testing this out with manual runs, the bug became apparent. For example, I triggered a manual run with `run_id=manual__2024-08-28T13:54:14.979062+00:00`. 

  - After over 25 runs, I tried opening this DAG run by clicking on the run_id in Browse-> DAG Runs. This opened the DAG without a specific run highlighted (see below).
  
  ![image](https://github.com/user-attachments/assets/161a4b5f-a802-4bee-854f-2abca8d6a1a9)

  -  When I clicked on a task in the graph view, it highlighted the run that had occurred just before the run I had clicked on in the grid. The `run_id` for this was `scheduled__2024-08-28T13:54:00+00:00` as indicated in the screenshot below. The run I wanted to open was not part of the grid. Hem observed this as well in his tests.
  
  ![image](https://github.com/user-attachments/assets/efedf055-a33c-4d4b-b708-9ab03f29a447)

  - In contrast, when I opened the DAG run for `run_id=scheduled__2024-08-28T13:54:00+00:00`, it opened and highlighted the correct DAG run. This might suggest that the bug only appears for manual runs, but this is not true as described in the next bullet point. 

- When I modified the DAG to have `schedule=@continuous`, the bug appeared once again for scheduled runs indicating that it is not exclusive to manual runs. For example, when I clicked on a run_id which was `scheduled__2024-08-28T16:47:22.632138+00:00`, I saw the same bug as before where it opened a page without any DAG run highlighted. When i clicked on one of the tasks in the graph, it highlighted the run previous to what I clicked on in the grid. The `run_id` for this was `scheduled__2024-08-28T16:47:20.415112+00:00`, and the run I was looking for was not visible in the grid. Hem observed this as well in his tests.

- Based on these observations, I believe the bug appears only for DAGs with run ID timestamps that are not at the 0th second of the minute. Due to the bug, you are only able to see the run previous to the run you wanted to locate. It does not depend on whether the run was scheduled or manually triggered. For example, the bug appeared for `run_id=scheduled__2024-08-28T16:47:22.632138+00:00` but not for `run_id=scheduled__2024-08-28T13:54:00+00:00`.
